src/types/final_data.rs

```diff
@@ -42,16 +42,104 @@ pub enum CompleteBound {
 
 /// Data related to a finalized type, which can be extracted from a [`super::Type`]
 /// if (and only if) it is finalized.
-#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
```

Comment:
9964c3ed8c229c85859f5acfbe7af164a9a83048: 4x speedup due to fewer `String` allocations and instant `Eq` and `Ord` computations (as opposed to recursive ones)?

Reply:
The 4x speedup is because we compare TMRs of complete bounds during unification, instead of re-recursing into their incomplete types.

Reply:
The `derive`s are removed because they're otherwise recursive, but there's not really any perf improvement (plausibly there would be a slight regression because of the extra indirection and heap allocations).

Reply:
I wrote an [alternative commit](https://github.com/uncomputable/rust-simplicity/commit/576af55ed1359bada7dcbb6611beb683dc607fc7) that only changes `Debug` and `Display`. The speedup is there.

Reply:
When I go one commit before the alternative commit, the tests are slow. After applying the commit, there is a speed up. I think the commit message was right. Anyway, not too important.

Reply:
Huh, yeah, I think you're right. I guess I was sloppy pinpointing where the speedup came from. (FWIW I only even noticed it several commits later, then used `git rebase -x` to run the tests on every commit, and see what the timing was. I was not trying to speed anything up in this PR.)

Agreed, it's not too important, so I won't change the message back, but I wish I hadn't changed it now.

Reply:
And looking at the code now, I think you're right that this is about the string allocations.